### PR TITLE
feat(cli): stdout option for cdk8s synth

### PR DESCRIPTION
**Pull request by campionfellin**
Closes https://github.com/awslabs/cdk8s/issues/92

Signed-off-by: campionfellin <campionfellin@gmail.com>

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 1]
```diff
@@ -12,11 +12,18 @@ class Command implements yargs.CommandModule {
 
   public readonly builder = (args: yargs.Argv) => args
     .option('app', { default: config.app, required: true, desc: 'Command to use in order to execute cdk8s app', alias: 'a' })
-    .option('output', { default: config.output, required: true, desc: 'Output directory', alias: 'o' });
+    .option('output', { default: config.output, required: false, desc: 'Output directory', alias: 'o' })
+    .option('stdout', { type: 'boolean', required: false, desc: 'Flag to print to stdout', alias: 'p' });
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:16]:
```suggestion
    .option('stdout', { type: 'boolean', required: false, desc: 'Print synthesized manifest to STDOUT', alias: 'p' });
```

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 2]
```diff
@@ -25,9 +32,14 @@ class Command implements yargs.CommandModule {
       env: {
         ...process.env,
         CDK8S_OUTDIR: outdir,
+        CDK8S_STDOUT: stdout,
       },
     });
 
+    if (stdout) {
+      process.exit(0);
+    }
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:41]: Why is this needed?

File: packages/cdk8s/src/app.ts [hunk 3]
```diff
@@ -9,11 +9,20 @@ import { Names } from './names';
...
    */
   readonly outdir?: string;
+
+  /**
+   * Whether or not to print to stdout. Mutually exclusive with
+   * the `--outdir` option.
+   *
+   * @default - true if defined, false otherwise
+   */
+  readonly stdout?: boolean;
```
Comment (eladb) [inline packages/cdk8s/src/app.ts:25]: mmmm.... not sure I am super comfortable with this implemented here. I am wondering if we should just implement this in the CLI by emitting the files to a temporary directory and then just `cat`ting those files to STDOUT. I don't know that this is something we really need in the framework. @iliapolo what do you think?

**Review by eladb — CHANGES_REQUESTED** [review]

File: packages/cdk8s-cli/src/cli/cmds/synth.ts (hunk 2 above)
Comment (campionfellin) [reply packages/cdk8s-cli/src/cli/cmds/synth.ts:41]: Skipping checks below this that verify an output file was created

File: packages/cdk8s/src/app.ts (hunk 3 above)
Comment (campionfellin) [reply packages/cdk8s/src/app.ts:25]: Yeah, I can move it to the CLI like that, works for me!

File: packages/cdk8s/src/app.ts [hunk 4]
```diff
@@ -9,7 +9,8 @@ import { Yaml } from './yaml';
 
 export interface AppOptions {
   /**
-   * The directory to output Kubernetes manifests.
+   * The directory to output Kubernetes manifests. Mutually exclusive with
+   * the `--stdout` option.
```
Comment (campionfellin) [inline packages/cdk8s/src/app.ts:13]:
```suggestion
   * The directory to output Kubernetes manifests.
```

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 5]
```diff
@@ -12,37 +12,43 @@ class Command implements yargs.CommandModule {
...
+    } else {
+      await shell(command, [], {
+        shell: true,
+        env: {
+          ...process.env,
+          CDK8S_OUTDIR: outdir,
+        },
+      });
+
+      validateSynthesis(outdir);
```
Comment (campionfellin) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:51]:
```suggestion
      await validateSynthesis(outdir);
```

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 6]
```diff
@@ -12,37 +12,43 @@ class Command implements yargs.CommandModule {
...
-      env: {
-        ...process.env,
-        CDK8S_OUTDIR: outdir,
-      },
-    });
-
-    if (!await fs.pathExists(outdir)) {
-      console.error(`ERROR: synthesis failed, app expected to create "${outdir}"`);
+    if (outdir !== config.output && stdout) {
+      console.error('ERROR: \'--output\' and \'--stdout\' are mutually exclusive. Please only use one.');
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:24]: I rather we throw an exception instead of `error+exit`.

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 7]
```diff
@@ -12,37 +12,43 @@ class Command implements yargs.CommandModule {
...
-      console.error('No manifests synthesized');
+    if (stdout) {
+      await mkdtemp(async tempDir => {
+        await shell(command, [], {
+          shell: true,
+          env: {
+            ...process.env,
+            CDK8S_OUTDIR: tempDir,
+          },
+        });
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:38]: Extract this to a function to avoid the duplication with lines 43...47

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 8]
```diff
@@ -12,37 +12,43 @@ class Command implements yargs.CommandModule {
...
+
+        await shell(`cat ${tempDir}/*.yaml`, [], { stdio: 'inherit', shell: true });
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:40]: Feels weird to shell out in order to read some files.

Something like this:

```ts
const manifests = (await fs.readdir(tempdir)).filter(f => path.extname(f) === '.k8s.yaml');
for (const f of manifests) {
  fs.createReadStream(path.join(tempdir, f)).pipe(process.stdout);
}
```

File: packages/cdk8s-cli/src/cli/cmds/synth.ts (hunk 7 above)
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:38]: Don't we want to call `validateSynthesis()` here as well?

**Review by eladb — CHANGES_REQUESTED** [review]

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 9]
```diff
@@ -12,37 +13,34 @@ class Command implements yargs.CommandModule {
...
+        await callLibrary(command, tempDir);
 
-    let found = false;
-    for (const file of await fs.readdir(outdir)) {
-      if (file.endsWith('.k8s.yaml')) {
-        console.log(`${outdir}/${file}`);
-        found = true;
-      }
-    }
+        const manifests = (await fs.readdir(tempDir)).filter(f => path.extname(f) === '.k8s.yaml');
```
Comment (campionfellin) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:34]:
```suggestion
        const manifests = (await fs.readdir(tempDir)).filter(f => path.extname(f) === '.yaml');
```

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 10]
```diff
@@ -12,37 +13,34 @@ class Command implements yargs.CommandModule {
...
-      if (file.endsWith('.k8s.yaml')) {
-        console.log(`${outdir}/${file}`);
-        found = true;
-      }
-    }
+        const manifests = (await fs.readdir(tempDir)).filter(f => path.extname(f) === '.yaml');
+        for (const f of manifests) {
+          fs.createReadStream(path.join(tempDir, f)).pipe(process.stdout);
+        }
+        await validateSynthesis(tempDir);
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:38]: Shouldn't this be done immediately (or as part of `callLibrary`)?

File: packages/cdk8s-cli/src/util.ts [hunk 11]
```diff
@@ -32,6 +32,35 @@ export async function mkdtemp(closure: (dir: string) => Promise<void>) {
   }
 }
 
+export async function callLibrary(command: string, outdir: string) {
```
Comment (eladb) [inline packages/cdk8s-cli/src/util.ts:35]: Rename to `synthApp` and combine with `validateSynthesis`

Comment (eladb) [reply packages/cdk8s-cli/src/util.ts:35]: Can you please also write a small unit test for this?

File: packages/cdk8s-cli/src/cli/cmds/synth.ts [hunk 12]
```diff
@@ -12,37 +13,34 @@ class Command implements yargs.CommandModule {
...
-        CDK8S_OUTDIR: outdir,
-      },
-    });
-
-    if (!await fs.pathExists(outdir)) {
-      console.error(`ERROR: synthesis failed, app expected to create "${outdir}"`);
-      process.exit(1);
-    }
+    if (stdout) {
+      await mkdtemp(async tempDir => {
```
Comment (eladb) [inline packages/cdk8s-cli/src/cli/cmds/synth.ts:31]: Would that remote the temporary directory after we are done with it?

**Review by eladb — CHANGES_REQUESTED** [review]

File: packages/cdk8s-cli/src/cli/cmds/synth.ts (hunk 12 above)
Comment (campionfellin) [reply packages/cdk8s-cli/src/cli/cmds/synth.ts:31]: Yep! This is what `mkdtemp` does:

```
  const workdir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s-'));
  try {
    await closure(workdir);
  } finally {
    await fs.remove(workdir);
  }
```

File: packages/cdk8s-cli/src/cli/cmds/synth.ts (hunk 10 above)
Comment (campionfellin) [reply packages/cdk8s-cli/src/cli/cmds/synth.ts]: Done!

File: packages/cdk8s-cli/src/util.ts (hunk 11 above)
Comment (campionfellin) [reply packages/cdk8s-cli/src/util.ts]: I wrote a couple of example tests, if you think that's the kind of style we want there I can fill them out a bit more!

**Review by eladb — APPROVED** [review]
nice and clean
